### DISCUSSION TITLE
What is the best way to validate multiple headers

### DISCUSSION THREAD

**Discussion by poul1x** [fastapi/fastapi#6902] [Questions] [answered]
### Problem

I decided to implement api versioning using header values: `api-version` (trunk changes), `resource-version` (leaf changes).  

I want to:
1) Ensure these two values are present in headers
2) Ensure these two values have `int` type
3) Use dependency injection to pass api version to endpoint

```Python
# Validation
class ApiVersionInfo(BaseModel):
    api_version: int
    resource_version: int

# Add dependency to app / router
def set_api_version_info(request: Request, api_ver: ApiVersionInfo = Header(...)):
    request.state.api_info = api_info

# Add dependency to endpoint
def get_api_version_info(request: Request):
    return request.state.api_info
```

But FastAPI does not allow to do this:

```
AssertionError: Param: api_info can only be a request body, using Body(...)
```

I've read https://fastapi.tiangolo.com/tutorial/header-params/.
But there's no common mechanism for checking and validation (like Query, Body, e.t.c).

**Question**: what is the best way to validate multiple headers?

**Comment by falkben** [comment, marked as answer]
If you specify the Header arguments individually in your dependency, you will get the validation that you want without having to create a pydantic model. In this situation, that is how I would I would go about it.

Your dependency would look like this:

```python
def set_api_version_info(
    request: Request, api_version: int = Header(...),  resource_version: int = Header(...)
):
    request.state.api_info = api_version
```

And then just use that as a dependency wherever you want it. It will do the validation you want.

If you're dead set on using a pydantic model, you need to overwrite the signature of the model to convert them to Headers.  A full example below, modeled after @Mause's excellent answer here about pydantic-> form: https://github.com/tiangolo/fastapi/issues/1989#issuecomment-684799715

```python
from fastapi import Depends, FastAPI, Header, Request
from pydantic import BaseModel


def as_header(cls):
    """decorator for pydantic model
    replaces the Signature of the parameters of the pydantic model with `Header`
    """
    cls.__signature__ = cls.__signature__.replace(
        parameters=[
            arg.replace(
                default=Header(...) if arg.default is arg.empty else Header(arg.default)
            )
            for arg in cls.__signature__.parameters.values()
        ]
    )
    return cls


@as_header
class ApiVersionInfo(BaseModel):
    api_version: int
    resource_version: int


def set_api_version_info(
    request: Request, api_ver: ApiVersionInfo = Depends(ApiVersionInfo)
):
    request.state.api_info = api_ver.api_version


app = FastAPI(dependencies=[Depends(set_api_version_info)])


@app.get("/")
def get_api_version_info(request: Request):
    return request.state.api_info
```

**Comment by poul1x** [comment]
> If you specify the Header arguments individually in your dependency, you will get the validation that you want without having to create a pydantic model.

@falkben, that's good solution, thanks!

**Comment by tiangolo** [comment]
Thanks for the help here @falkben ! 👏  🙇 

Thanks for reporting back and closing the issue @poul1x  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
